File: lesson7giang/src/main/java/com/example/lesson7giang/StudentViewModel.java
package com.example.lesson7giang;

import android.app.Application;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Observer;

import java.util.List;

public class StudentViewModel extends AndroidViewModel {
    private StudentRepository mRepository;
    LiveData<List<Student>> mAllStudent;

    public StudentViewModel(@NonNull Application application) {
        super(application);
        mRepository = new StudentRepository(application);
        mAllStudent = mRepository.getAllStudent();
    }


    LiveData<List<Student>> getAllStudent(){
        return  mAllStudent;
    }

    public void insert(Student student){
        mRepository.insert(student);
    }

    public int size(){
        return mAllStudent.getValue().size();
    }

    public String getName(int position){
        return mAllStudent.getValue().get(position).getName();
    }

    public void deleteAll(){
        mAllStudent.removeObserver((Observer<? super List<Student>>) mAllStudent);
    }
}

File: lesson7khanh/src/main/java/com/k001/lesson7khanh/StudentViewModel.kt
package com.k001.lesson7khanh

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class StudentViewModel(application: Application): AndroidViewModel(application) {

    private val repository: StudentRepository
    val allStudent: LiveData<List<Student>>

    init{
        val studentDao = StudentRoomDatabase.getDatabase(application, viewModelScope).studentDao()
        repository = StudentRepository(studentDao)
        allStudent = repository.allStudent
    }

    fun insert(student: Student) = viewModelScope.launch(Dispatchers.IO){
        repository.insert(student)
    }

    fun delete(student: Student) = viewModelScope.launch(Dispatchers.IO){
        repository.delete(student)
    }
}